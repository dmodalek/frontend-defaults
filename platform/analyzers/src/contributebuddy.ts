import { directoryExists, getDependencyInstallation, DependencyInstallation } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type ContributebuddyAnalyzerResult = {
    contributebuddy: boolean;
    contributebuddyConfigFolder?: string,
    contributebuddyInstallation?: DependencyInstallation
};

export const contributebuddyAnalyzer = async (cwd: string): Promise<ContributebuddyAnalyzerResult> => {
    const CONTRIBUTE_BUDDY_FOLDER = join(cwd, '.contributebuddy');
    const contributebuddyConfigFolder = await directoryExists(CONTRIBUTE_BUDDY_FOLDER);
    const contributebuddyInstallation = await getDependencyInstallation(cwd, 'contribute-buddy');

    if (contributebuddyConfigFolder || contributebuddyInstallation.declared) {
        return {
            contributebuddy: true,
            contributebuddyConfigFolder: CONTRIBUTE_BUDDY_FOLDER,
            contributebuddyInstallation

        }
    }

    return {
        contributebuddy: false,
    };
}