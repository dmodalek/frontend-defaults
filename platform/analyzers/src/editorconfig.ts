import { Analyzer, fileExists } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type EditorconfigAnalyzerResult = {
	editorConfig: boolean;
};

export class EditorconfigAnalyzer extends Analyzer<EditorconfigAnalyzerResult> {
	async analyze(): Promise<EditorconfigAnalyzerResult> {
		const eidtorConfigExists = await fileExists(join(this.context, '.editorconfig'));

		return {
			editorConfig: eidtorConfigExists,
		};
	}
}
