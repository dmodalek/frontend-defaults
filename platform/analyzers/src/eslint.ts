import { DependencyInstallation, fileExists, getDependencyInstallation } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type ESLintAnalyzerResult = {
	eslint: boolean;
	eslintIgnore?: boolean;
	eslintInstallation?: DependencyInstallation
};

export const eslintAnalyzer = async (cwd: string): Promise<ESLintAnalyzerResult> => {
	const doesESLintExist = await fileExists(join(cwd, '.eslintrc.js'));

	if (doesESLintExist) {
		const doesESLintIgnoreExist = await fileExists(join(cwd, '.eslintignore'));
		const eslintInstallation = await getDependencyInstallation(cwd, 'eslint');

		return {
			eslint: doesESLintExist,
			eslintIgnore: doesESLintIgnoreExist,
			eslintInstallation
		};
	}

	return {
		eslint: doesESLintExist,
	};
}