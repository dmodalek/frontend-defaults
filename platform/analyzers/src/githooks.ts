import { join } from 'path';
import { Analyzer, fileExists } from '@namics/frontend-defaults-platform-core';

export type GitHooksLocation = 'rc' | 'package';

export type GitHooksAnalyzerResult = {
	githooks: boolean;
	githooksInstallation?: false | string;
	githooksLocation?: GitHooksLocation;
};

export class GitHooksAnalyzer extends Analyzer<GitHooksAnalyzerResult> {
	async analyze(): Promise<GitHooksAnalyzerResult> {
		const doesGitHooksConfigExist = await fileExists(join(this.context, '.huskyrc'));
		const isHuskyInstalled = this.packageAnalyzer.anyDependencyExists('husky');
		const doesGitHooksConfigExistInPkg = !!(this.package as any).husky;

		if (doesGitHooksConfigExist || doesGitHooksConfigExistInPkg) {
			return {
				githooks: doesGitHooksConfigExist,
				githooksLocation: doesGitHooksConfigExistInPkg ? 'package' : 'rc',
				githooksInstallation: isHuskyInstalled ? isHuskyInstalled.version : false,
			};
		}

		return {
			githooks: false,
		};
	}
}
