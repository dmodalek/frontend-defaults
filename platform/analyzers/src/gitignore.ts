import { join } from 'path';
import { Analyzer, fileExists } from '@namics/frontend-defaults-platform-core';

export type GitIgnoreAnalyzerResult = {
	gitignore: boolean;
};

export class GitIgnoreAnalzyer extends Analyzer<GitIgnoreAnalyzerResult> {
	async analyze(): Promise<GitIgnoreAnalyzerResult> {
		const doesGitignoreExist = await fileExists(join(this.context, '.gitignore'));

		return {
			gitignore: doesGitignoreExist,
		};
	}
}
