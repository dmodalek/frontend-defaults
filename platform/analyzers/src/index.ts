// export all related analyzer stuff for contributebuddy
export { contributebuddyAnalyzer, ContributebuddyAnalyzerResult } from './contributebuddy';

// export all related analyzer stuff for editorconfig
export { editorConfigAnalyzer, EditorconfigAnalyzerResult } from './editorconfig';

// export all related analyzer stuff for eslint
export { eslintAnalyzer, ESLintAnalyzerResult } from './eslint';

// export all related analyzer stuff for githooks
export { gitHooksAnalyzer, GitHooksAnalyzerResult } from './githooks';

// export all related analyzer stuff for gitignore
export { gitIgnoreAnalyzer, GitIgnoreAnalyzerResult } from './gitignore';

// export all related analyzer stuff for license
export { licenseAnalyzer, LicenseAnalyzerResult } from './license';

// export all related analyzer stuff for nitro
export { nitroAnalyzer, NitroAnalyzerResult } from './nitro';

// export all related analyzer stuff for node
export { nodeAnalyzer, NodeAnalyzerResult } from './node';

// export all related analyzer stuff for npmrc
export { npmrcAnalyzer, NPMRCAnalyzerResult } from './npmrc';

// export all related analyzer stuff for prettier
export { prettierAnalyzer, PrettierAnalyzerResult } from './prettier';

// export all related analyzer stuff for readme
export { readmeAnalyzer, ReadmeAnalyzerResult } from './readme';

// export all related analyzer stuff for repository
export { repositoryAnalyzer, RepositoryAnalyzerResult } from './repository';

// export all related analyzer stuff for stylelint
export { stylelintAnalyzer, StylelintAnalyzerResult } from './stylelint';

// export all related analyzer stuff for tslint
export { tslintAnalyzer, TSLintAnalyzerResult } from './tslint';

// export all related analyzer stuff for typescript
export { typescriptAnalyzer, TypeScriptAnalyzerResult } from './typescript';

// export all related analyzer stuff for webpack
export { webpackAnalyzer, WebpackAnalyzerResult } from './webpack';
