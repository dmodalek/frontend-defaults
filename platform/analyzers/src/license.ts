import { Analyzer, fileExists } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type LicenseAnalyzerResult = {
	license: boolean;
};

export class LicenseAnalyzer extends Analyzer<LicenseAnalyzerResult> {
	async analyze(): Promise<LicenseAnalyzerResult> {
		const licenseExists = await fileExists(join(this.context, 'LICENSE'));

		return {
			license: licenseExists,
		};
	}
}
