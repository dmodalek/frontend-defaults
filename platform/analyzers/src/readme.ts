import { Analyzer, fileExists } from '@namics/frontend-defaults-platform-core';
import { join } from 'path';

export type ReadmeAnalyzerResult = {
	readme: boolean;
};

export class ReadmeAnalyzer extends Analyzer<ReadmeAnalyzerResult> {
	async analyze(): Promise<ReadmeAnalyzerResult> {
		const readmeExists = await fileExists(join(this.context, 'README.md'));

		return {
			readme: readmeExists,
		};
	}
}
