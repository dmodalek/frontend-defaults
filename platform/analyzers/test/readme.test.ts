import { ProjectAnalyzer } from '@namics/frontend-defaults-platform-core';
import { getFixturePath } from './suite';
import { ReadmeAnalyzerResult, ReadmeAnalyzer } from '../src/readme';

const FIXTURE = getFixturePath('readme-project');
const FIXTURE_NO_README = getFixturePath('default-project');

const getFixtureAnalyzer = async (context: string): Promise<ProjectAnalyzer<ReadmeAnalyzerResult>> => {
	return new ProjectAnalyzer<ReadmeAnalyzerResult>({
		context,
		analyzers: [ReadmeAnalyzer],
	}).boot();
};

describe('Analyzers', () => {
	describe('ReadmeAnalyzer', () => {
		it('should not crash the ProjectAnalyzer', async () => {
			expect(async () => await getFixtureAnalyzer(FIXTURE)).not.toThrow();
			expect(async () => await getFixtureAnalyzer(FIXTURE_NO_README)).not.toThrow();
		});

		it('should analyze a project with readme correctly', async () => {
			const analyzer = await getFixtureAnalyzer(FIXTURE);

			expect(analyzer.analytics.readme).toEqual(true);
			expect(analyzer.analytics).toMatchSnapshot();
		});

		it('should analyze a project without readme correctly', async () => {
			const analyzer = await getFixtureAnalyzer(FIXTURE_NO_README);

			expect(analyzer.analytics.readme).toEqual(false);
			expect(analyzer.analytics).toMatchSnapshot();
		});
	});
});
