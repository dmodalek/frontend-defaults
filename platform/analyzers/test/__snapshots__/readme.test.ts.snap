// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Analyzers ReadmeAnalyzer should analyze a project with readme correctly 1`] = `
Object {
  "readme": true,
}
`;

exports[`Analyzers ReadmeAnalyzer should analyze a project without readme correctly 1`] = `
Object {
  "readme": false,
}
`;
